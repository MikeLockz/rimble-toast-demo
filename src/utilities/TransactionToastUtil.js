import React from "react";

import { ToastMessage } from "rimble-ui";

class TransactionToastUtil extends React.Component {
  // Determines if collections are same size
  collectionHasNewObject = (prevCollection, currentCollection) => {
    return (
      typeof prevCollection === "undefined" ||
      Object.keys(prevCollection).length !==
        Object.keys(currentCollection).length
    );
  };

  // Returns object from currentCollection that doesn't exist in prevCollection
  getNewObjectFromCollection = (prevCollection, currentCollection) => {
    if (typeof prevCollection !== "undefined") {
      const objectKey = Object.keys(currentCollection).filter(key => {
        return !Object.keys(prevCollection).includes(key);
      });
      return currentCollection[objectKey];
    } else {
      return Object.keys(currentCollection).map(key => {
        return currentCollection[key];
      });
    }
  };

  // Compare two collections of objects, return single object from current collection that differs from prev collection
  getUpdatedObjectFromCollection = (prevCollection, currentCollection) => {
    const updatedTransaction = Object.keys(prevCollection)
      .map(key => {
        if (
          prevCollection[key].lastUpdated !== currentCollection[key].lastUpdated
        ) {
          return currentCollection[key];
        }
      })
      .filter(object => typeof object !== "undefined");
    return updatedTransaction[0];
  };

  // Returns an object from a collection based on a given identifier
  getObjectFromCollection = (identifier, collection) => {
    const object = collection[`tx${identifier}`];
    return object;
  };

  // Returns either a new object or finds an updated object in a collection against a previous collection
  getUpdatedTransaction = (prevCollection, currentCollection) => {
    let tx = null;
    let currentTx = {};
    let prevTx = {};
    
    if (this.collectionHasNewObject(prevCollection, currentCollection)) {
      tx = this.getNewObjectFromCollection(prevCollection, currentCollection);
    } else {
      currentTx = this.getUpdatedObjectFromCollection(
        prevCollection,
        currentCollection
      );
      prevTx = this.getObjectFromCollection(currentTx.created, prevCollection);

      if (currentTx.status !== prevTx.status) {
        tx = currentTx;
      }
    }
    return tx;
  };

  // Check for updates to the transactions collection
  processTransactionUpdates = prevProps => {
    
    let tx = null;
    if (Object.keys(this.props.transactions).length) {
      tx = this.getUpdatedTransaction(
        prevProps.transactions,
        this.props.transactions
      );
    }

    if (tx !== null && typeof tx !== "undefined") {
      this.showTransactionToast(tx);
    }
  };

  showTransactionToast = transaction => {
    // Get text info for toast
    let toastMeta = this.getTransactionToastMeta(transaction);

    // Show toast
    window.toastProvider.addMessage(".", toastMeta);
  };

  getTransactionToastMeta = transaction => {
    let transactionToastMeta = {};
    let status = transaction.status;
    let transactionHash = transaction.transactionHash;

    switch (status) {
      case "started":
        transactionToastMeta = {
          message: "Change submitted",
          secondaryMessage: "Confirm in MetaMask",
          actionHref: "",
          actionText: "",
          variant: "default",
          icon: "InfoOutline"
        };
        break;
      case "pending":
        transactionToastMeta = {
          message: "Processing change...",
          secondaryMessage: "This may take a few minutes",
          actionHref: "",
          actionText: "",
          variant: "processing"
        };
        break;
      case "confirmed":
        transactionToastMeta = {
          message: "First block confirmed",
          secondaryMessage: "Your change is in progress",
          actionHref: "https://rinkeby.etherscan.io/tx/" + transactionHash,
          actionText: "Check progress",
          variant: "processing"
        };
        break;
      case "success":
        transactionToastMeta = {
          message: "Smart contract value changed",
          variant: "success"
        };
        break;
      case "error":
        transactionToastMeta = {
          message: "Value change failed",
          secondaryMessage:
            "Make sure you have enough Ether (ETH) and try again",
          actionHref: "",
          actionText: "",
          variant: "failure"
        };
        break;
      default:
        break;
    }
    return transactionToastMeta;
  };

  componentDidUpdate(prevProps, prevState) {
    this.processTransactionUpdates(prevProps);
  }

  render() {
    return (
      <div>
        <ToastMessage.Provider ref={node => (window.toastProvider = node)} />
      </div>
    );
  }
}

export default TransactionToastUtil;
