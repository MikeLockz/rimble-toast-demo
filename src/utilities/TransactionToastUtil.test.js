import React from "react";
import { shallow } from "enzyme";
import TransactionToastUtil from "./TransactionToastUtil";

describe("TransactionToastUtil", () => {
  const componentProps = {
    tx1551107396364: {
      created: 1551107396364,
      lastUpdated: 1551107430741,
      status: "success",
      confirmationCount: 24,
      method: "incrementCounter",
      type: "contract",
      transactionHash:
        "0x18c56dc171eb8e6f95d0489c8e18a8fde95ee1b8350a624abaa17d2872894f41",
      recentEvent: "receipt"
    },
    tx1551107423018: {
      created: 1551107423018,
      lastUpdated: 1551107462155,
      status: "success",
      confirmationCount: 24,
      method: "incrementCounter",
      type: "contract",
      transactionHash:
        "0xfe92a2e59ef090df835b48a3795eb4f80e8673b0ce1ce3f4aa060dcbc6a26f7a",
      recentEvent: "receipt"
    },
    tx1551107428376: {
      created: 1551107428376,
      lastUpdated: 1551107462064,
      status: "success",
      confirmationCount: 24,
      method: "incrementCounter",
      type: "contract",
      transactionHash:
        "0xb3f476d68a7148de593bf758d9a449ffbb2da994f14ae0e97695872432d94908",
      recentEvent: "receipt"
    },
    tx1551107627514: {
      created: 1551107627514,
      lastUpdated: 1551107631111,
      status: "pending",
      confirmationCount: 0,
      method: "incrementCounter",
      type: "contract"
    }
  };

  const updatedProps = {
    tx1551107396364: {
      created: 1551107396364,
      lastUpdated: 1551107430741,
      status: "success",
      confirmationCount: 24,
      method: "incrementCounter",
      type: "contract",
      transactionHash:
        "0x18c56dc171eb8e6f95d0489c8e18a8fde95ee1b8350a624abaa17d2872894f41",
      recentEvent: "receipt"
    },
    tx1551107423018: {
      created: 1551107423018,
      lastUpdated: 1551107462155,
      status: "success",
      confirmationCount: 24,
      method: "incrementCounter",
      type: "contract",
      transactionHash:
        "0xfe92a2e59ef090df835b48a3795eb4f80e8673b0ce1ce3f4aa060dcbc6a26f7a",
      recentEvent: "receipt"
    },
    tx1551107428376: {
      created: 1551107428376,
      lastUpdated: 1551107462064,
      status: "success",
      confirmationCount: 24,
      method: "incrementCounter",
      type: "contract",
      transactionHash:
        "0xb3f476d68a7148de593bf758d9a449ffbb2da994f14ae0e97695872432d94908",
      recentEvent: "receipt"
    },
    tx1551107627514: {
      created: 1551107627514,
      lastUpdated: 1551107639999,
      status: "pending",
      confirmationCount: 0,
      method: "incrementCounter",
      type: "contract"
    }
  };

  it("renders children", () => {
    const wrapper = shallow(<TransactionToastUtil />);
    expect(wrapper).toMatchSnapshot();
  });
});
