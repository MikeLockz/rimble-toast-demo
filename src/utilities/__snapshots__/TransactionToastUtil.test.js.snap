// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionToastUtil renders children 1`] = `
<div>
  <Toast Message Provider
    delay={3000}
    messageData={
      Object {
        "actionHref": "",
        "actionText": "",
        "icon": "",
        "message": "[…Placeholder Message…]",
        "secondaryMessage": "",
        "variant": "default",
      }
    }
  />
</div>
`;
